kv: don't unquiesce uninitialized replicas

nvb opened the pull request:
In a [support issue](https://github.com/cockroachlabs/support/issues/1340), we saw that 10s of thousands of uninitialized replicas were being ticked regularly and creating a large amount of background work on a node, driving up CPU. This commit updates the Raft quiescence logic to disallow uninitialized replicas from being unquiesced and Tick()'ing themselves.

Keeping uninitialized replicas quiesced even in the presence of Raft traffic avoids wasted work. We could Tick() these replicas, but doing so is unnecessary because uninitialized replicas can never win elections, so there is no reason for them to ever call an election. In fact, uninitialized replicas do not even know who their peers are, so there would be no way for them to call an election or for them to send any other non-reactive message. As a result, all work performed by an uninitialized replica is reactive and in response to incoming messages (see `processRequestQueue`).

There are multiple ways for an uninitialized replica to be created and then abandoned, and we don't do a good job garbage collecting them at a later point (see https://github.com/cockroachdb/cockroach/issues/73424), so it is important that they are cheap. Keeping them quiesced instead of letting them unquiesce and tick every 200ms indefinitely avoids a meaningful amount of periodic work for each uninitialized replica.

Release notes (bug fix): uninitialized replicas that are abandoned after an unsuccessful snapshot no longer perform periodic background work, so they no longer have a non-negligible cost.

nvb committed:
kv: access HLC clock once per Replica tick

This is a small refactor that eliminates a call to `HLC.Now` on each replica
tick. Due to the issue fixed by the following commit, we were ticking 1000s of
uninitialized replicas in fast succession and as a result, these clock accesses
were a major source of mutex contention. We shouldn't be ticking uninitialized
replicas, but since we now know that this is an expensive part of Replica.tick,
we might as well make it cheaper.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73362)
<!-- Reviewable:end -->

nvanbenschoten commented:
> Perhaps it is important that the uninitialized replica calls an election, if only so that it gets told to clean itself up if it has been abandoned. Except that this might be broken, because uninitialized replicas are learners, so they aren't even promotable and so they won't call elections. I think I'm missing something here.

I think this is incorrect in two ways which offset. Uninitialized replicas don't know that they are learners, because they don't have access to the range descriptor. This means that they aren't disqualified for being "promotable" because of that, at least from their perspective. However, uninitialized replicas don't know the other members of their range and don't even track themselves in their `Progress` map, so they still [aren't "promotable"](https://github.com/etcd-io/etcd/blob/7572a61a39d4eaad596ab8d9364f7df9a84ff4a3/raft/raft.go#L1620) and wouldn't even know who to talk to if they were and decided to call an election. So ticking an uninitialized replica doesn't do anything.

This also means that leader elections cannot be used to trigger the cleanup of uninitialized replicas.

@tbg's new state machine diagram of replica initialization also hints at there being no path for an uninitialized replica to be removed without going through an initialized replica.

![image (5)](https://user-images.githubusercontent.com/5438456/144363620-2d81a3d7-fa18-4f59-a99c-60c007e25385.png)

nvanbenschoten commented:
https://github.com/cockroachdb/cockroach/pull/47982 looks relevant. I guess that confirms the suspicion.

So I guess the next steps here are:
1. polishing this PR off (i.e. with a test), now that we're more confident that ticking uninitialized replicas is entirely useless. This makes uninitialized replicas cheaper and less disruptive.
2. landing https://github.com/cockroachdb/cockroach/pull/47982 so that we will eventually (after ~1h) GC these abandoned uninitialized replicas.

@ajwerner do you have interest in pushing that PR over the finish line?

tbg commented:
Let's keep Andrew out of it, I'm sure he has enough to do.

I might be missing something, but isn't GC'ing uninitialized replicas difficult? We would need an index RangeID->RangeDesc which we don't have (meta ranges are keyed on EndKey, not RangeID) or do full meta scans every now and then.
We can't "just" remove uninitialized replicas on a hunch without having proved that they are no longer a member of the range as this amounts to reducing availability of the range in a way that is more permanent than having a replica wait for a snapshot (removing a voter and preventing it from being recreated). We can remove non-voters (and expect that to be the common case) but to confirm that we need to see the descriptor, too.

Also worth pausing to think about the benefit of replicaGC'ing uninit'ed replicas once we don't tick them. Once we no longer tick them, the main upshot is removing them from the in-memory maps. The disk state is negligibly different, RangeTombstone key vs HardState key.

tbg commented:
We can definitely replicaGC uninitialized replicas that have a trivial HardState, i.e. in particular `Vote` is unset. That should be the majority of cases. But what if `Vote` isn't unset? We know this uninitialized replica was a voter and we strongly suspect that it's obsolete, but there could also be permanent problems getting a snapshot there and losing this voter might lose quorum in a pretty permanent way.

tbg reviewed (approved):
> Reviewed 4 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73362)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[-- commits, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuKyb66cwo6-MSHGqe):*
It's by design in the sense that we can't replicaGC them for the reasons discussed in the main thread. There's definitely no desire to not be able to replicaGC them but it hasn't been a priority.
___
*[-- commits, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuL7kT4owB88QdtU7Q):*
Correct, which might never happen, for example if the range doesn't exist any more (merge).
___
*[-- commits, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuLFWFBYep3oYenx_L):*
nit: they aren't always learners. You can also get uninit'ed voters with a quick split-rebalance succession. But the uninitialized replica doesn't even know what it is, it will have an empty config, see here: https://github.com/cockroachdb/cockroach/blob/4c8c0d42933921e1ac821a9bdcde5c2059f2a97a/pkg/kv/kvserver/replica_raftstorage.go#L66-L69

So they certainly will not call an election, ever, since the progress map is empty. They also can't send a message to anyone in general, since that would require them to at least know about a peer, but these would have to be in the configuration. The only thing they can do is respond to incoming messages.

This indicates that by not ticking uninit'ed replicas, we're not regressing on their ability to "commonly" replicaGC themselves in response to a rejection of one of their outgoing messages, as this basically never happens in the first place anyway.
___
*[pkg/kv/kvserver/store_raft.go, line 638 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuMf8C4LF1nfHYPEhW:-MpuMf8C4LF1nfHYPEhX:bhs9gug) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a7631ba3d6d597af1fc50c838cf3142ca0be18b/pkg/kv/kvserver/store_raft.go#L638)):*
> ```Go
> 
> 			// Filter out uninitialized replicas. We could tick them, but doing so is
> 			// unnecessary because uninitialized replicas can never win elections, so
> ```

Might want to update this to reflect our finding that uninited replicas cannot even send non-reactive messages since they don't know who the peers are.
___
*[pkg/kv/kvserver/store_raft.go, line 647 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuN3px0j_ASvwqPKdN:-MpuN3py89gcghsEq7QJ:b-v9w3qq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a7631ba3d6d597af1fc50c838cf3142ca0be18b/pkg/kv/kvserver/store_raft.go#L647)):*
> ```Go
> 			// them through the Raft scheduler and ticking them avoids a meaningful
> 			// amount of periodic work for each uninitialized replica.
> 			s.mu.RLock()
> ```

Reading this makes me wonder why we allow uninited replicas to be unquiesced. Not suggesting you change that, but maybe add a comment saying that it's just the way it is and there is no good reason and we don't like lots of things about uninitialized replicas and see https://github.com/cockroachdb/cockroach/issues/72374.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> isn't GC'ing uninitialized replicas difficult? We would need an index RangeID->RangeDesc which we don't have (meta ranges are keyed on EndKey, not RangeID) or do full meta scans every now and then.

This makes sense to me, thanks for explaining. This must mean that the approach in https://github.com/cockroachdb/cockroach/pull/47982 is broken. It seems to perform its consistent meta lookup with the replica's in-memory RangeDescriptor [here](https://github.com/cockroachdb/cockroach/pull/47982/files#diff-6b507a68a64ff356be424284b60be855af1958c61281957432c22bccede895daR290). But if the replica is uninitialized, this descriptor will be empty. So it won't actually look up the right range.

> Also worth pausing to think about the benefit of replicaGC'ing uninit'ed replicas once we don't tick them. Once we no longer tick them, other than a sense of closure the main upshot is removing them from the in-memory maps. The disk state is negligibly different, RangeTombstone key vs HardState key.

:thumbsup: that matches my understanding as well.

> I lean towards doing a ~daily scan through the meta ranges per node and performing GC of uninitialized replicas based on that, since that is simple and effective and handles the problem fully.

This seems nice and simple.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-:-MpwJb6d-bcKlA7kSv4i:bz8bjdg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/kv/kvserver/store_raft.go, line 647 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuN3px0j_ASvwqPKdN:-MpwLJeO7dfFdqcarhza:b1eo0fj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a7631ba3d6d597af1fc50c838cf3142ca0be18b/pkg/kv/kvserver/store_raft.go#L647)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Reading this makes me wonder why we allow uninited replicas to be unquiesced. Not suggesting you change that, but maybe add a comment saying that it's just the way it is and there is no good reason and we don't like lots of things about uninitialized replicas and see https://github.com/cockroachdb/cockroach/issues/72374.
</blockquote></details>

That's a really interesting idea. If we allowed uninited replicas to quiesce (on their own, with no leader input) then we wouldn't need any of this special handling. These replicas would be removed from the `unquiescedReplicas` map and stay out of it until they heard from a peer. I'll explore that alternate approach, as it seems pretty clean and conceptually consistent.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
Thanks for all your work on this @nvanbenschoten! Wrote up an issue to handle GC in #73424.

nvb committed:
kv: clean up unquiescedReplicas insertion in addReplicaToRangeMapLocked

Refactor that does not change behavior.

nvb committed:
kv: ignore internalRaftGroup in Replica.Metric

A replica cannot unquiesce if its internalRaftGroup is nil, so this
condition was redundant but confusing.

nvb committed:
kv: don't unquiesce uninitialized replicas

In a [support issue](https://github.com/cockroachlabs/support/issues/1340), we
saw that 10s of thousands of uninitialized replicas were being ticked regularly
and creating a large amount of background work on a node, driving up CPU. This
commit updates the Raft quiescence logic to disallow uninitialized replicas from
being unquiesced and Tick()'ing themselves.

Keeping uninitialized replicas quiesced even in the presence of Raft traffic
avoids wasted work. We could Tick() these replicas, but doing so is unnecessary
because uninitialized replicas can never win elections, so there is no reason
for them to ever call an election. In fact, uninitialized replicas do not even
know who their peers are, so there would be no way for them to call an election
or for them to send any other non-reactive message. As a result, all work
performed by an uninitialized replica is reactive and in response to incoming
messages (see processRequestQueue).

There are multiple ways for an uninitialized replica to be created and
then abandoned, and we don't do a good job garbage collecting them at a
later point (see https://github.com/cockroachdb/cockroach/issues/73424),
so it is important that they are cheap. Keeping them quiesced instead of
letting them unquiesce and tick every 200ms indefinitely avoids a
meaningful amount of periodic work for each uninitialized replica.

Release notes (bug fix): uninitialized replicas that are abandoned after an
unsuccessful snapshot no longer perform periodic background work, so they no
longer have a non-negligible cost.

nvanbenschoten commented:
Thanks for opening the issue @erikgrinaker! I updated the code comment to point to that issue.

@tbg I rewrote this PR according to your suggestion. Instead of letting uninitialized replicas unquiesce but then filtering them out in the tick loop, we now no longer let uninitialized replicas unquiesce. This seems cleaner to me, as it introduces a reasonable relationship between state machine initialization and quiescence and uses the existing quiescence logic to prevent these replicas from ticking, instead of introducing an additional orthogonal condition for ticking. PTAL.

tbg reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 2 of 4 files at r4, 9 of 9 files at r9, 4 of 4 files at r10, 4 of 4 files at r11, 4 of 4 files at r12, 7 of 7 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-:-MqJuHYv4HtsndV-nPdv:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/client_raft_test.go, line 4105 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJsCkMBs3QsguyN6-n:-MqJsCkMBs3QsguyN6-o:b-pt6g8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4105)):*
> ```Go
> 	defer tc.Stopper().Stop(ctx)
> 
> 	_, desc, err := tc.Servers[0].ScratchRangeEx()
> ```

Add a comment about using an expiration-based lease or, my preference if possible, switch to `ScratchRange`.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4110 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrb6N10zwJMhR9Hrz:-MqJrb6O-42HihE--DwV:b-j7iu43) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4110)):*
> ```Go
> 	require.NoError(t, tc.WaitForSplitAndInitialization(key))
> 
> 	// Block a snapshot.
> ```

// Block incoming snapshots on s2 until channel is signaled.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4114 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrPj3-mcQaagHsugU:-MqJrPj3-mcQaagHsugV:bhc77x1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4114)):*
> ```Go
> 	handlerFuncs := noopRaftHandlerFuncs()
> 	handlerFuncs.snapErr = func(header *kvserver.SnapshotRequest_Header) error {
> 		<-blockSnapshot
> ```

give up on `stopper.ShouldQuiesce()`?
___
*[pkg/kv/kvserver/client_raft_test.go, line 4117 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrl2b-Hu88dgSYBY5:-MqJrl2b-Hu88dgSYBY6:b-84tpx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4117)):*
> ```Go
> 		return nil
> 	}
> 	s1, err := tc.Server(1).GetStores().(*kvserver.Stores).GetStore(tc.Server(1).GetFirstStoreID())
> ```

s2? s1 is on `tc.Server(0)`. Indexing, always a joy.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4125 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrhiI6OXeLY8SPku_:-MqJrhiI6OXeLY8SPkua:b-a0hks5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4125)):*
> ```Go
> 	})
> 
> 	// Try to up-replicate to s1. Should block on a learner snapshot after the new
> ```

s2 here and below?
___
*[pkg/kv/kvserver/client_raft_test.go, line 4130 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrqLY979QqhhgYhz2:-MqJrqLY979QqhhgYhz3:b-to2441) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4130)):*
> ```Go
> 	// receiving Raft traffic from the leader.
> 	replicateErrChan := make(chan error)
> 	go func() {
> ```

nit: I know it doesn't matter here but if we have a stopper might as well use it.
___
*[pkg/kv/kvserver/replica.go, line 306 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJsPPA5_2vbR7olV6G:-MqJsPPA5_2vbR7olV6H:b-az77d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica.go#L306)):*
> ```Go
> 		//
> 		// Replica objects always begin life in a quiescent state, as the field is
> 		// set to true in the Replica constructor newUnloadedReplica. They unquiesce
> ```

Ohh, interesting. I had no idea. That explains your earlier commit. I take back my comment about ` || internalRaftGroup == nil` then.
___
*[pkg/kv/kvserver/replica.go, line 310 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJs_sT0_BaF_fRQ-za:-MqJs_sT0_BaF_fRQ-zb:b-kguqek) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica.go#L310)):*
> ```Go
> 		// unquiesceWithOptionsLocked, which are called in response to Raft traffic.
> 		//
> 		// Only initialized replicas that have a non-nil internalRaftGroup are
> ```

This makes it sound like there are initialized replicas with a nil raft group, which raises the question of how these are allowed to unquiesce (since they are not according to this comment). So either init'ed replicas with nil group don't exist, or whether the group is nil or not doesn't matter for purposes of unquiescing?
___
*[pkg/kv/kvserver/replica_metrics.go, line 62 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJqi4S5L_xx3e6SH2N:-MqJqi4S5L_xx3e6SH2O:bhdtyff) ([raw file](https://github.com/cockroachdb/cockroach/blob/d466b09cb34e46cc70834afab44a45ae006df169/pkg/kv/kvserver/replica_metrics.go#L62)):*
> ```Go
> 	raftStatus := r.raftStatusRLocked()
> 	leaseStatus := r.leaseStatusAtRLocked(ctx, now)
> 	quiescent := r.mu.quiescent || r.mu.internalRaftGroup == nil
> ```

Interesting. Have I missed a commit? Because, wouldn't an uninitialized replica start out with `quiescent = false` now when it was previously `true`? As is, this seems like a behavior change to me. Not a terribly important one (since this is only for `ReplicaMetrics` and `internalRaftGroup` is populated anyway within ~10 minutes (scan interval)) but still, this doesn't quite check out with the commit message.
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 52 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJsuzS41lhIjQ2mSr1:-MqJsuzS41lhIjQ2mSr2:b-uenymu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica_raft_quiesce.go#L52)):*
> ```Go
> }
> 
> func (r *Replica) unquiesceWithOptionsLocked(campaignOnWake bool) {
> ```

`maybeUnquiesceWithOptionsLocked` to indicate that it may not be possible?
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 116 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJtKO65TH7xZP6PpIV:-MqJtKO65TH7xZP6PpIW:b7x9w2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica_raft_quiesce.go#L116)):*
> ```Go
> 		// lazily in response to any Raft traffic (see stepRaftGroup) or KV request
> 		// traffic (see maybeInitializeRaftGroup). If it has yet to be initialized,
> 		// let it remain quiesced. The Raft group will be initialized soon enough.
> ```

This is pretty annoying; we morally say that a replica unquiesces when it needs to use raft, but then we also say this isn't true, you need some folks bypassing this check and creating the group first? This is pretty complex and I'm not at all clear on why this works out in practice (I assume we have the right calls in the actual critical path).

I think we should initialize the raft group right away. Is now perhaps the right time to make that change? Doing this lazily was a half-baked way to avoid election storms back when this was still a problem and possibly before we even had quiescence. Now that quiescence is the default, what's the benefit of the lazy raft group?

I understand that we're looking to backport this, so maybe a fast-follow on master only?
___
*[pkg/kv/kvserver/store_create_replica.go, line 351 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJuao96UWh2_Xy1rgN:-MqJuao96UWh2_Xy1rgO:b-vs6gqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/store_create_replica.go#L351)):*
> ```Go
> 	// quiesced, so we don't need to campaign or wake the leader. We just want
> 	// to start ticking.
> 	lockedRepl.unquiesceWithOptionsLocked(false /* campaignOnWake */)
> ```

Beating the same drum as before, but now we have to know that the internal raft group is no longer nil (probably because we hit `withReplicaForRequest`?)

Did you have to add this line in response to any failures? It makes sense, but shouldn't be necessary, right?


<!-- Sent from Reviewable.io -->

nvb committed:
kv: rename unquiesceXYZ methods to maybeUnquiesceXYZ

This commit renames the `unquiesceXYZ` methods to include a "maybe" prefix to
indicate that they may not succeed in unquiescing the replica.

The commit also introduces a boolean return value to these methods to inform
the caller about whether the unquiesce attempt was successful.

Finally, the commit uses this return value in `maybeMarkReplicaInitializedLockedReplLocked`
to assert that after initializing a replica, it did unquiesce immediately.

nvb reviewed (commented):
TFTR! I finally got around to addressing your comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-:-MqeSWbcCPmLBzIm_K4S:b-12qtzj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/client_raft_test.go, line 4105 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJsCkMBs3QsguyN6-n:-MqeRQiwDuU71Z4pMH9T:b-cm1cxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4105)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Add a comment about using an expiration-based lease or, my preference if possible, switch to `ScratchRange`.
</blockquote></details>

`ScratchRangeEx` isn't an expiration-based lease, it's the "extended" version of `ScratchRange` that returns the left and right descriptor instead of a start key. This made more sense to me than the logic I see in some places where we `ScatchRange` and then take the returned start key a look up the corresponding descriptor that we just dropped in `ScratchRange`.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4110 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrb6N10zwJMhR9Hrz:-MqeRwkP6eaj1OBnxd9w:b-fabz67) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4110)):*
> // Block incoming snapshots on s2 until channel is signaled.
Done.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4114 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrPj3-mcQaagHsugU:-MqeRyPI7WeIZxKkSH8Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4114)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

give up on `stopper.ShouldQuiesce()`?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4117 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrl2b-Hu88dgSYBY5:-MqeS9Bb597OtkUL_BuM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4117)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

s2? s1 is on `tc.Server(0)`. Indexing, always a joy.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4125 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrhiI6OXeLY8SPku_:-MqeS9xM6D2g1DxHxxX7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4125)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

s2 here and below?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/client_raft_test.go, line 4130 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJrqLY979QqhhgYhz2:-MqeSQiz3FygDWVHCitp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/client_raft_test.go#L4130)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: I know it doesn't matter here but if we have a stopper might as well use it.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica.go, line 306 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJsPPA5_2vbR7olV6G:-MqeSk6K-KYY13LcDvHw:bhq3kqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica.go#L306)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ohh, interesting. I had no idea. That explains your earlier commit. I take back my comment about ` || internalRaftGroup == nil` then.
</blockquote></details>

Me neither until writing this up. But I think it makes sense for things to work like that.
___
*[pkg/kv/kvserver/replica.go, line 310 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJs_sT0_BaF_fRQ-za:-Mqe_qHE40sPCiXpdNQq:buzfehd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica.go#L310)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This makes it sound like there are initialized replicas with a nil raft group, which raises the question of how these are allowed to unquiesce (since they are not according to this comment). So either init'ed replicas with nil group don't exist, or whether the group is nil or not doesn't matter for purposes of unquiescing?
</blockquote></details>

The dependency is the other way around. There are initialized replicas with a nil raft group. They unquiesce by first initializing their raft group in `withRaftGroupLocked` and then trying to unquiesce. But let's get rid of this craziness in https://github.com/cockroachdb/cockroach/issues/73715.
___
*[pkg/kv/kvserver/replica_metrics.go, line 62 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJqi4S5L_xx3e6SH2N:-MqeTD8J8N28PwLCioRo:b-1nbd4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/d466b09cb34e46cc70834afab44a45ae006df169/pkg/kv/kvserver/replica_metrics.go#L62)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Interesting. Have I missed a commit? Because, wouldn't an uninitialized replica start out with `quiescent = false` now when it was previously `true`? As is, this seems like a behavior change to me. Not a terribly important one (since this is only for `ReplicaMetrics` and `internalRaftGroup` is populated anyway within ~10 minutes (scan interval)) but still, this doesn't quite check out with the commit message.
</blockquote></details>

I think this comment came before your "I take back my comment", but in case it wasn't clear, all replicas (initialized or uninitialized) start with `quiescent = true`. They can then only unquiesce if `r.mu.internalRaftGroup != nil`, which has always been the case and is not changing in this PR. So `!r.mu.quiescent` implies `r.mu.internalRaftGroup != nil`, meaning that this was redundant.
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 52 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJsuzS41lhIjQ2mSr1:-MqeaPD57neOlk_rMw47:b-tpqhhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica_raft_quiesce.go#L52)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`maybeUnquiesceWithOptionsLocked` to indicate that it may not be possible?
</blockquote></details>

Done. I also added a return value that we can use for assertions in a few places. PTAL at the new commit.
___
*[pkg/kv/kvserver/replica_raft_quiesce.go, line 116 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJtKO65TH7xZP6PpIV:-MqeYheh67ej2u-gSb8F:b-4nf4lr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/replica_raft_quiesce.go#L116)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is pretty annoying; we morally say that a replica unquiesces when it needs to use raft, but then we also say this isn't true, you need some folks bypassing this check and creating the group first? This is pretty complex and I'm not at all clear on why this works out in practice (I assume we have the right calls in the actual critical path).

I think we should initialize the raft group right away. Is now perhaps the right time to make that change? Doing this lazily was a half-baked way to avoid election storms back when this was still a problem and possibly before we even had quiescence. Now that quiescence is the default, what's the benefit of the lazy raft group?

I understand that we're looking to backport this, so maybe a fast-follow on master only?
</blockquote></details>

I think I see what you're saying. The idea would be that if quiescence is the default and we only unquiesce in all the same places that would otherwise initialize the raft group, then this lazy raft group creation isn't doing anything anymore, and so the dependency from `internalRaftGroup != nil -> !quiescent` is unnecessary complexity. I think I agree with that in theory, as it simplifies the states that a replica can be in significantly.

But I'd like to avoid expanding the scope of this PR any further, so a fast-follow sounds good. I opened https://github.com/cockroachdb/cockroach/issues/73715 to track the work.
___
*[pkg/kv/kvserver/store_create_replica.go, line 351 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MqJuao96UWh2_Xy1rgN:-MqeU2sbDtQ7pRzAppSw:bqc20kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b6639249ad8df99528886126e1d7c03a25981ae/pkg/kv/kvserver/store_create_replica.go#L351)):*
> but now we have to know that the internal raft group is no longer nil (probably because we hit withReplicaForRequest?)

Yes, that's correct. If we made it here then it was in response to a raft message and so the internal raft group is no longer nil.

> Did you have to add this line in response to any failures? It makes sense, but shouldn't be necessary, right?

No, this wasn't in response to any failures. In fact, it's not strictly necessary because this method is only ever called inside of `handleRaftReadyRaftMuLocked` in response to snapshots, which also [calls `withRaftGroupLocked` with a `true` return value for `unquiesceAndWakeLeader` before completing](https://github.com/cockroachdb/cockroach/blob/406e0fe1e44cabc8a3384703c642469567c6915e/pkg/kv/kvserver/replica_raft.go#L916). So if we ever initialize ourselves inside of `handleRaftReadyRaftMuLocked` and don't unquiesce before getting to the end, that will take care of unquiescing for us.

But this isn't entirely superfluous — there is a benefit to calling this here instead of waiting. That's because doing so allows us to call `unquiesceWithOptionsLocked` and to pass `!campaignOnWake`. So since we know we don't need to campaign or wake the leader, it's better to unquiesce now and prevent the other code path from doing so.

I added both of these points to the comment.
___
*[pkg/kv/kvserver/store_raft.go, line 638 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-MpuMf8C4LF1nfHYPEhW:-MqeTzsF-DlXsVaX0xA2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a7631ba3d6d597af1fc50c838cf3142ca0be18b/pkg/kv/kvserver/store_raft.go#L638)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Might want to update this to reflect our finding that uninited replicas cannot even send non-reactive messages since they don't know who the peers are.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 8 of 8 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, 6 of 6 files at r17, 10 of 10 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73362)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/store_create_replica.go, line 369 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-Mqs4_Yj-34fIvwiI7kv:-Mqs4_Yj-34fIvwiI7kw:b-9his43) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4f814dea5a04f1a7b91919a7e61d0aca07a6eb7/pkg/kv/kvserver/store_create_replica.go#L369)):*
> ```Go
> 	// the range, so we can unquiesce without campaigning or waking the leader.
> 	if !lockedRepl.maybeUnquiesceWithOptionsLocked(false /* campaignOnWake */) {
> 		return errors.Errorf("expected replica %s to unquiesce after initialization", desc)
> ```

`errors.AssertionFailedf`? If we hit this the world will end, right? Can't imagine this will be handled gracefully by the callers.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTR!

bors r=tbg

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-:-MqvBKwZ5w29pDyS8KaB:b-kg5fnc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvserver/store_create_replica.go, line 369 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/73362#-Mqs4_Yj-34fIvwiI7kv:-MqvB1N_7qpnG6kPLkH5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4f814dea5a04f1a7b91919a7e61d0aca07a6eb7/pkg/kv/kvserver/store_create_replica.go#L369)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`errors.AssertionFailedf`? If we hit this the world will end, right? Can't imagine this will be handled gracefully by the callers.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
